### PR TITLE
Windows: Add generate-swagger-api.ps1

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard <jhoward@microsoft.com>

@johnstep Adds a Windows native way of doing this without the need to run under WSL or a Linux VM.

### CODE REVIEW

**Review by dnephin — COMMENTED** [review]
Could we modify the original script to work in windows as well?  I thought windows had bash now? These are identical except for the top 2 lines

**Comment by jhowardmsft** [comment]
@dnephin - WSL is an optional component, it's not built-in at all.

File: hack/generate-swagger-api.ps1 [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    -n ImageSummary `
+    -n Plugin -n PluginDevice -n PluginMount -n PluginEnv -n PluginInterfaceType `
+    -n Port `
+    -n ServiceUpdateResponse `
+    -n Volume
+
+swagger generate operation -f api/swagger.yaml `
+    -t api -a types -m types -C api/swagger-gen.yaml `
+    -T api/templates --skip-responses --skip-parameters --skip-validator `
+    -n Authenticate `
```
Comment (thaJeztah) [inline hack/generate-swagger-api.ps1]: Can we add a comment here, and in the Linux counterpart that when updating this list, the corresponding file for the other platform must be updated as well?

**Review by thaJeztah — CHANGES_REQUESTED** [review]
LGTM

File: hack/generate-swagger-api.ps1 (hunk 1 above)
Comment (jhowardmsft) [reply hack/generate-swagger-api.ps1]: @thaJeztah Ooops. Totally missed that comment. Done now.

**Comment by thaJeztah** [comment]
ping @johnstep @dnephin PTAL

**Comment by dnephin** [comment]
I still don't like duplicating this. Bash works on windows, we should use the bash script.

**Comment by kolyshkin** [comment]
So, what we can do is
```bash
git mv hack/generate-swagger-api.sh hack/generate-swagger-api.ps1
ln -s generate-swagger-api.ps1 hack/generate-swagger-api.sh
```
and change the first two lines of the script to this one:

`#!/bin/sh -eu`

If the assumption that Windows PowerShell ignores shebang is correct, this will work.

**Comment by jhowardmsft** [comment]
@dnephin WSL is an optional subsystem. @kolyshkin Except that symlinks in cloned repos on Windows don't work well at all...

**Comment by thaJeztah** [comment]
@jhowardmsft the thought was to do it the other way round; rename the `generate-swagger-api.sh` script to `generate-swagger-api.ps1`, then making `generate-swagger-api.sh` a symlink to `generate-swagger-api.ps1`.

That way cloning the repo on _Linux_ uses a symlink, but on Windows, it uses the actual file

**Comment by jhowardmsft** [comment]
Yeah, but there will still be a symlink.....

**Comment by dnephin** [comment]
Let's just remove the `.sh` extension , then we don't need a symlink.

**Comment by thaJeztah** [comment]
ping @jhowardmsft ^^ will that do the job?

**Comment by jhowardmsft** [comment]
No it still won't work, although the shebang update is fine - it's just a comment and not interpreted. The issues are that the continuation character in PS is a backtick rather than backslash, so it balks (and AFAIK it's not configurable); The second is that AFAIK, short of some convoluted workarounds such as ```Invoke-Command -ScriptBlock ([ScriptBlock]::Create((Get-Content .\generate-swagger-api)))```, it's difficult to invoke PowerShell to run a script which doesn't have a `.ps1` extension. Unless @johnstep knows something I don't here.

**Comment by thaJeztah** [comment]
@jhowardmsft with WSL being available everywhere, is this one still important to have? Discussing this one again in the maintainers meeting, but the concerns remain about the scripts getting out of sync

**Comment by olljanat** [comment]
Derek add label: rebuild/*

**Comment by olljanat** [comment]
Derek add label: rebuild/z

**Comment by lowenna** [comment]
Long out-of-date. Closing as I am no longer working in this area.

**Pull request closed without merging**